### ISSUE TITLE
Layout of enumerations

### ISSUE THREAD

**Issue by jklmnn** [Componolit/ada-style#24]
Single line (for completeness):
```Ada
type Enum is (A, B, C) with Condition;
```

A:
```Ada
type Enum is (CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_A => 1,
              CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_B => 2,
              CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_C => 3) with
   Aspect => Condition;
```
A2 (if A does not fit column size, used in the Ada Style Guide, p22):
```Ada
type Enum is
   (CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_A => 1,
    CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_B => 2,
    CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_C => 3) with
Aspect => Condition;
```
A3:
```Ada
type Enum is
   (CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_A => 1,
    CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_B => 2,
    CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_C => 3)
with Aspect => Condition;
```
B:
```Ada
type Enum is (
   CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_A => 1,
   CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_B => 2,
   CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_C => 3
) with Aspect => Condition;
```
C:
```Ada
type Enum is
(
   CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_A => 1,
   CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_B => 2,
   CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_C => 3
) with Aspect => Condition;
```
C2:
```Ada
type Enum is
(
   CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_A => 1,
   CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_B => 2,
   CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_C => 3
)
with Aspect => Condition;
```

**Comment by senier** [comment]
@jklmnn CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS? The existing specs may give some ideas.

**Comment by treiher** [comment]
I would prefer a clear separation of the literals from the type name and the aspects. That is not given for **A**, **A2** and **C** (because of the `with` keyword). I don't like that in **A** the indentation of the literals depends on the type name. This leads to a different indentation for each enumeration. It could also lead to space problems for long literals.

```Ada
type Enum_With_Very_Long_Name is (CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_A => 1,
                                  CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_B => 2,
                                  CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_C => 3) with
   Aspect => Condition;

type Enum is (A => 1,
              B => 2,
              C => 3) with
   Aspect => Condition;

type Enum_With_Long_Name is (D => 1,
                             E => 2,
                             F => 3) with
   Aspect => Condition;
```

In **C** having the opening parenthesis on a separate line, but not the closing parenthesis looks inconsistent. I have added a small variation **C2** in which this is changed, but the result is a bit bulky.

My current favorites are **B** and **C2**.

**Comment by jklmnn** [comment]
**C2** feels to me like a lot unnecessarily wasted space with separate lines just for the parentheses. I see the problem with the mixed value/aspect one line. I added **A3** to see how it looks with the `with` on a separate line.
Just as a side node, A and A2 would be consistent with our current use of `with` in subprograms and records.
Apart from that I think **B** would be a good compromise in terms of used space and having values and aspects on different lines.

**Comment by senier** [comment]
I agree with @treiher's statement on **A**, **A2** and **C**. While I like the symmetry of **C2**, I agree that this variant is a bit wasteful wrt. to vertical space. I'm undecided between **A3** and **B**. The dangling open bracket in **B** IMHO visually is somewhat unpleasant. **A3** would be consistent with subprogram parameter lists.

**Comment by treiher** [comment]
I think **A3** would be also a good option. **B** and **A3** seem to be the most liked variants among the three of us.

> Just as a side node, A and A2 would be consistent with our current use of `with` in subprograms and records.

You are right. Actually, I don't like the current use of `with` in subprograms either. But I think it makes sense to have a consistent style for similar things. **A** would be consistent to the current style:

```Ada
type Enum is (CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_A => 1,
              CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_B => 2,
              CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_C => 3) with
   Aspect => Condition;

procedure Foo (A : Natural;
               B : Natural) with
  Pre  => True,
  Post => True
is
   <declarations>
begin
   <statements>
end Bar;
```

The consistent use of **B** would mean the following:

```Ada
type Enum is (
   CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_A => 1,
   CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_B => 2,
   CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_C => 3
) with Aspect => Condition;

procedure Foo (
   A : Natural;
   B : Natural
) with 
   Pre  => True,
   Post => True
is
   <declarations>
begin
   <statements>
end Bar;
```

The consistent use of **A3** would look as follows:

```Ada
type Enum is
   (CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_A => 1,
    CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_B => 2,
    CAN_YOU_CHANGE_THE_EXAMPLES_TO_CONTAIN_SOME_LONGER_LITERALS_C => 3)
with Aspect => Condition;

procedure Foo 
   (A : Natural;
    B : Natural) 
with
   Pre  => True,
   Post => True
is
   <declarations>
begin
   <statements>
end Bar;
```

From the three variants **A3** would clearly be my favorite. (To be completely consistent, there would have to be a line break after the `with` keyword also for the enumerations.)

**Comment by senier** [comment]
I also favor **A3** now (I'd leave the aspect on the same line as the `with` - to be completely consistent with subprograms we'd also have to put the `is` on a separate line, which I don't think would be helpful).

@jklmnn , please add **A3** to the style guide.

**Comment by jklmnn** [comment]
I noticed we made a mistake. We have to discuss the style for Ada and RecordFlux partially different since the syntax differs. The syntax used in the examples is not valid Ada. https://github.com/Componolit/ada-style/pull/25#discussion_r564462507 shows this difference. I would say that we should allow
```Ada
type Enum is (A, B, C) with Aspect => Condition;
```
in Ada but not
```
type Enum is (A => 1, B => 2, C => 3) with Aspect => Condition;
```
in RecordFlux. Although IMHO the latter should not be defined here, since it is not part of Ada.

**Comment by senier** [comment]
Good catch! Also, we should state that enumeration literals are always upper case.

**Comment by jklmnn** [comment]
> Good catch! Also, we should state that enumeration literals are always upper case.

In general or in RecordFlux?

**Comment by treiher** [comment]
I think we should keep the style as much as possible consistent between Ada and RecordFlux.

**Comment by jklmnn** [comment]
While I'm usually all for consistency, I don't think that too much of it is helpful here. This would mean that would have to change
```Ada
type Enum is (A, B, C);
```
to
```Ada
type Enum is
   (A,
    B,
    C);
```
which I do not find helpful.
On the other hand due to the required values
```Ada
type Enum is
   (A => 1,
    B => 2,
    C => 3) with Aspect => Condition;
```
is much more readable than
```Ada
type Enum is (A => 1, B => 2, C => 3) with Aspect => Condition;
```
I think we should attribute for the different requirements in the style. Also note that at least the size aspect is required in RecordFlux increasing the line length further while enums in Ada don't have aspects in the most cases.

**Comment by treiher** [comment]
I get your point. Then this is probably the wrong place to document our decision. I think RecordFlux's language reference can be used to document the RecordFlux-specific coding style.

**Comment by treiher** [comment]
Would the following be the result for Ada?

```Ada
type Enum is (A, B, C);

type Enum_Short is (A, B, C) with Aspect => Condition;

type Enum_Long is
   (Very_Very_Very_Very_Long_Literal_1,
    Very_Very_Very_Very_Long_Literal_2,
    Very_Very_Very_Very_Long_Literal_3)
with Aspect => Condition;
```

**Comment by jklmnn** [comment]
> Would the following be the result for Ada?
> 
> ```ada
> type Enum is (A, B, C);
> 
> type Enum_Short is (A, B, C) with Aspect => Condition;
> 
> type Enum_Long is
>    (Very_Very_Very_Very_Long_Literal_1,
>     Very_Very_Very_Very_Long_Literal_2,
>     Very_Very_Very_Very_Long_Literal_3)
> with Aspect => Condition;
> ```

Yes. Unless there aren't any objections I would adapt the pull request to add this.

**Comment by jklmnn** [comment]
Fixed by #25.
